### PR TITLE
[Android, iOS] Fixes layout compression causes

### LINKED ISSUE #3624: Layout Compression causes the app to crash when scrolling a ListView with ListViewCachingStrategy.RetainElement
### Description
Scrolling through a `ListView` whose `ItemTemplate` contains elements with `CompressedLayout.IsHeadless="true"` results in the app crashing. Note that this seems to happen only when the caching strategy is set to `RetainElement` (working fine with recycle).

`System.InvalidCastException: Specified cast is not valid.` is thrown on iOS and `[MonoDroid] System.NullReferenceException: Object reference not set to an instance of an object.` on Android, but the stack traces look related.

iOS:
`Unhandled Exception:
System.InvalidCastException: Specified cast is not valid.
  at (wrapper castclass) System.Object.__castclass_with_cache(object,intptr,intptr)
  at Xamarin.Forms.Platform.iOS.VisualElementRenderer 1[TElement].Xamarin.Forms.Platform.iOS.IVisualElementRenderer.SetElement (Xamarin.Forms.VisualElement element) [0x00000] in <2a59efab866341818ab4748ebe270f0a>:0 
  at Xamarin.Forms.Platform.iOS.RendererPool.UpdateRenderers (Xamarin.Forms.Element newElement) [0x0008c] in <2a59efab866341818ab4748ebe270f0a>:0 
  at Xamarin.Forms.Platform.iOS.RendererPool.UpdateNewElement (Xamarin.Forms.VisualElement newElement) [0x00080] in <2a59efab866341818ab4748ebe270f0a>:0 ....`

Android:
`[MonoDroid] UNHANDLED EXCEPTION:
[MonoDroid] System.NullReferenceException: Object reference not set to an instance of an object.
[MonoDroid]   at Xamarin.Forms.Platform.Android.VisualElementPackager.SetElement (Xamarin.Forms.VisualElement oldElement, Xamarin.Forms.VisualElement newElement) [0x00136] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementPackager.cs:264 
[MonoDroid]   at Xamarin.Forms.Platform.Android.VisualElementPackager.<.ctor>b__10_0 (System.Object sender, Xamarin.Forms.Platform.Android.VisualElementChangedEventArgs args) [0x00000] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementPackager.cs:39 
[MonoDroid]   at Xamarin.Forms.Platform.Android.VisualElementRenderer 1[TElement].OnElementChanged ...`
### Steps to Reproduce

1. Create a `ListView` and keep the caching strategy as `RetainElement`
2. Add a `StackLayout` to the `ItemTemplate` and set `CompressedLayout.IsHeadless="true"`
3. Add a number of items to the `ItemsSource` (enough for the list to scroll)
4. Scrolling the list crashes the app

- Version with issue: 3.1.0.697729
- Last known good version: -
- IDE:
- Platform Target Frameworks:
  - iOS:  11.4
  - Android: 8.1
- Affected Devices: multiple (e.g. : iPhone X, Nexus 6)

### Reproduction Link

[LayoutCompressionCrash.zip](https://github.com/xamarin/Xamarin.Forms/files/2306251/LayoutCompressionCrash.zip)

### PR DESCRIPTION
**Pull request by paymicro**
### Description of Change ###

Fixes layout compression causes in Android and iOS.

### Issues Resolved ### 

- fixes #3624

### API Changes ###
 
 None

### Platforms Affected ### 

- iOS
- Android

### Behavioral/Visual Changes ###

None

### Before/After Screenshots ### 

Not applicable

### Testing Procedure ###
- Run UITest 3624
- Click on button "With Compression (type 1) --- Crash while scrolling"
- Scroll page. It should not throw an exception

### PR Checklist ###

- [ ] Has automated tests 
- [x] Rebased on top of the target branch at time of PR
- [x] Changes adhere to coding standard

### CODE REVIEW

File: Xamarin.Forms.Platform.iOS/RendererPool.cs [hunk 1]
```diff
@@ -148,7 +148,7 @@ void UpdateRenderers(Element newElement)
 
 				var x = (int)childRenderer.NativeView.Layer.ZPosition / 1000;
 				var element = newElementController.LogicalChildren[x] as VisualElement;
-				if (element == null)
+				if (element == null || childRenderer.Element.GetType() != element.GetType())
```
Comment (rmarinho) [inline Xamarin.Forms.Platform.iOS/RendererPool.cs]: Why do we need this? is this related with CompressLayout? isn't this check being handle here https://github.com/xamarin/Xamarin.Forms/blob/a7a90885a9f0557988d4c5373717bb48b917dd09/Xamarin.Forms.Platform.iOS/RendererPool.cs#L53

Comment (paymicro) [reply Xamarin.Forms.Platform.iOS/RendererPool.cs]: Yes. This is best done in the place where this variable is defined. Fixed.

**Comment by rmarinho** [comment]
build --uitests

**Comment by rmarinho** [comment]
Build is failing i think you are missing a namespace.

**Comment by PureWeen** [comment]
build --uitests
